clients/drcachesim/tests/tlb_simulator_unit_test.cpp

```diff
@@ -0,0 +1,187 @@
+/* **********************************************************
+ * Copyright (c) 2015-2023 Google, LLC  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, LLC OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+#include "tlb_simulator_unit_test.h"
+#include "../simulator/tlb_simulator.h"
+#include "../common/memref.h"
+#include "trace_entry.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+class tlb_simulator_mock_t : public tlb_simulator_t {
+public:
+    tlb_simulator_mock_t(const tlb_simulator_knobs_t &knobs)
+        : tlb_simulator_t(knobs) {};
+
+    bool
+    process_memref(const memref_t &memref)
+    {
+        // Process memref like tlb_simulator does.
+        bool result = tlb_simulator_t::process_memref(memref);
+        // Save address used by tlb_simulator, if any.
+        if (!type_has_address(memref.data.type)) {
+            return result;
+        }
+        const memref_t *simref = &memref;
+        memref_t phys_memref;
+        if (knobs_.use_physical) {
+            phys_memref = memref2phys(memref);
+            simref = &phys_memref;
+        }
+        addresses.insert(simref->data.addr);
```

Comment:
(Hmm, so for the virtual address case, this isn't really testing much: you create a vector of memrefs from a list of addresses and then here you create a new set of addresses from the memrefs.)

Reply:
Yeah, I guess it checks `memref2phys()` does not have any effect in this case (or, more precisely, it checks that it's successfully skipped). This `check_addresses()` test was more for physical addresses to check for a correct translation, checking "the absence of a translation" when using virtual addresses was basically free, so that's why I added it. Should I just remove it?

Reply:
No, may as well leave it.

Reply:
Sounds good.